april_tag_updates/docking_script.py

```diff
@@ -0,0 +1,127 @@
+visited = []
+timeout =  10
+angle_zero = 10
+step1_done = False
+step2_done = False
+step3_done = False 
+step4_done = False
+motor = BasicMotorController(True)
+start_time = None
+
+while True:
+    if start_time is None:
+        start_time = time.time()
+    ret, image = cap.read()
+    (corners, ids, rejected) = cv2.aruco.detectMarkers(image, arucoDict, parameters=arucoParams)
+    if len(corners) == 0:
+      if time.time() - start_time > timeout and step1_done is False:
+        #Step 1 - Check for april tags rn by rotating in circle for 15 seconds.
+          motor.turn_right()
+      if step1_done and step2_done is False:
+        #Step 2 - go back until len(corners) = 1 or when we see at least 1 april tag
+        motor.reverse()
+    if len(corners) > 0:
+      if step1_done:
+        step2_done = True
+      start_time = time.time()
+      for (corner, markerId) in zip(corners, ids):
+          markerId = float(markerId)
+          (topLeft, topRight, bottomRight, bottomLeft) = corner[0]
+          pixel_width = np.sqrt((bottomRight[0] - bottomLeft[0])**2 + (bottomRight[1] - bottomLeft[1])**2)
+          depth = distance_to_camera(known_width,focal_length,pixel_width)
+          cxt,cyt = int((topLeft[0] + bottomRight[0])//2),int((topLeft[1] + bottomRight[1])//2)
+          cx, cy = int(960),int(540)
+          cv2.circle(image, (cxt, cyt), radius = 20, color = (0, 0, 255), thickness = -1)
+          cv2.circle(image, (cx, cy), radius = 20, color = (0, 0, 255), thickness = -1)
+          cv2.line(image,(cxt, cyt),(cx, cy), color = (0, 0, 0), thickness = 2)
+          ratio= known_width/pixel_width
+          dpix = np.sqrt((cx-cxt)**2+(cy-cyt)**2)
+          horiz_dist = dpix * ratio
+          midpoint = (int((cx + cxt) //2 ), int((cy + cyt) //2 - 25 ))
+          midpoint2 = (int((cx + cxt) //2), int((cy + cyt) //2 + 50 ))
+          midpoint3 = (int((cx + cxt) //2), int((cy + cyt) //2 - 100 ))
+          cv2.putText(image, f"{horiz_dist: .2f} in", midpoint, cv2.FONT_HERSHEY_COMPLEX, 1, (0, 101, 255), 2)
+          angle = np.degrees(np.arcsin(horiz_dist/(depth)))
+          cv2.putText(image, f"{angle:.2f} degrees", midpoint2, cv2.FONT_HERSHEY_COMPLEX, 1, (0, 101, 255), 2)
+          cv2.putText(image, f"{depth:.2f} in (depth)", midpoint3, cv2.FONT_HERSHEY_COMPLEX, 1, (255,0, 255), 2)
+          #Step 1 starts here
+          if idDict.get(markerId) == "R2": ##TODO: Add B2, L2, F2 in the future
+            #Step 1 is over. Start Step 2: move forward until tag out of sight. "
+            if angle <= angle_zero:
+              motor.go_forward()
+              step1_done = True
+            else:
+              # Step 1 - "Keep rotating till we reach center of middle tag"
+              if direction_of_tag(corner[0],image) == "right":
+                motor.turn_left()
+              else: 
+                motor.turn_right()
+          else:
+            #All of this is Step 1 
```

Comment:
step 1 or step 2?

Reply:
it's both - they are interweaved

Reply:
Initially it was step 2 but now we just changed it to be part of step 1.

Reply:
ok. where's step 2?

Reply:
if you to the top of the script, look under if len(corners) < 0 and under that look under if step1_done and not step2_done, step 2 starts there. Step 2 ends under if len(corners) > 0 and if step1_done: then step2_done.

Reply:
I see.It might be helpful to remove the comment on step 1 ending here then?